### ISSUE TITLE
morphology.reconstruction crashes python

### ISSUE THREAD

**Issue by horsto** [scikit-image/scikit-image#4078]
## Description
I encounter examples where running morphology.reconstruction() crashes python without error message - it just throws a python system error in windows after which the kernel has to be restarted. The input arrays do not seem to be particular in any way, and not getting any error message makes it really hard for me to see where things go wrong. This is only happening for specific arrays and - surprisingly - stochastically after 1-5 runs of this command.

## Way to reproduce
```python
se = morphology.disk(1)
Ie = morphology.erosion(map_, se)
fmap = morphology.reconstruction(Ie, map_)
```
where `map_` has shape (32, 32) is dtype('float64') and contains nans.

## Version information
```python
3.6.9 |Anaconda, Inc.| (default, Jul 30 2019, 13:42:17) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]
Darwin-18.6.0-x86_64-i386-64bit
scikit-image version: 0.15.0
numpy version: 1.16.4
```

**Comment by alexandrejaguar** [comment]
Hi @horsto,
I can reproduce it. IMHO, I think morphology algorithms are not made to deal with images containing `nan`, but I'll ask for my @scikit-image/core fellows to present their opinions too.
I would replace `nan` in this case with the minima of the image (always clarifying the process :)), but I'm not sure if or how they are important to you. Could you elaborate on that? Thanks!

**Comment by hmaarrfk** [comment]
Generally, the policy is that the software should not crash.

Whether they give meaningful results, is a different question

**Comment by sciunto** [comment]
Full code to reproduce


`
import numpy as np
from skimage import morphology
map_ = np.zeros((32, 32), dtype=np.float64)
map_[0, 0] = np.nan
se = morphology.disk(1)
Ie = morphology.erosion(map_, se)
fmap = morphology.reconstruction(Ie, map_)
`

`skimage/morphology/greyreconstruct.py:127: RuntimeWarning: invalid value encountered in greater
  if method == 'dilation' and np.any(seed > mask):
zsh: segmentation fault (core dumped)  ipython
`

**Comment by horsto** [comment]
Thanks a lot for all your replies. 

@alexandrejaguar: To be honest I lack intuition for how nans would influence the result compared to replacement by minima as you suggest. But I just tested it and it seems totally fine - so going with that for now. No more errors/crashes since then. 
@hmaarrfk: Agree, maybe nans (if that's really the problem) should be handled better?

**Comment by hmaarrfk** [comment]
two things generally happen when nans are involed:

1. A true edge case is found.
2. Inequalities don't have the same meaning for `nans`

<details>

```python
In [1]: import numpy as np

In [2]: np.nan
Out[2]: nan

In [3]: np.nan > 1
Out[3]: False

In [4]: np.nan < 1
Out[4]: False

In [5]: np.nan <= 1
Out[5]: False

In [6]: np.nan >= 1
Out[6]: False

In [7]: np.nan >= np.nan
Out[7]: False

In [8]: np.nan == np.nan
Out[8]: False
```
</details>

The code is taken from cellprofiler, maybe we can look there to see how they fixed it?

**Comment by jni** [comment]
Yeah the inequalities all failing could lead to an unexpected code path being taken that ends up accessing out-of-bounds memory. A quick fix is to add a mask (the terminology is confusing with this function, I mean a True/False mask of which pixels will get examined) of `np.isfinite` with the data. But probably the logic of the function should instead be cleaned up so that out of bounds accesses are impossible.

**Comment by lagru** [comment]
I agree with most of what was said here. I think how NaN's ought to be handled is highly specific to the algorithm and its implementation (which is a bad thing). E.g. using a mask for the local_maxima function wouldn't be a fix, would it? 

I would support measures that make the behavior more transparent:

- Address NaN's in the documentation if arrays of floats must be accepted and advise the user how to handle this (mask, how to replace, interpret results). E.g. it was handled with a general warning in the docstring of SciPy's [find_peaks](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html) function. 
- I'd find a runtime warning most intuitive but always checking for NaN's may not be acceptable performance wise.
- In any case a function should never crash and either raise a ValueError, warning or proceed as normal.

**Comment by hmaarrfk** [comment]
I think the addition of calls like isfinite is fine, we should add a keyword for it. Disabled by default.

But the out of bounds access should also be addressed.

**Comment by hmaarrfk** [comment]
Its definitely an out of bound access. This patch refactors the code to make the out of bounds access obvious

(I mean, there are other issues there too, uint32_t is used while int is meant to be used, or maybe even `intp`)

Can't really spend more time on this, but maybe somebody else wants to take it over from here.

memoryview is an easy way to create a null context.

[outof_bound.diff.txt](https://github.com/scikit-image/scikit-image/files/3465787/outof_bound.diff.txt)

**Comment by gitanupam** [comment]
For those who land here, morphology.reconstruction was silently crashing for me as well. I realised the size of the list/array was high (3 levels, number of elements totalling to ~17 million) which 1GB RAM on GCP could not handle. Increasing the RAM fixed it
